pip_api/_parse_requirements.py

```diff
@@ -463,8 +476,36 @@ def parse_requirements(
         lines_enum = _skip_regex(lines_enum, options)
 
         for lineno, line in lines_enum:
-            req: Optional[Union[requirements.Requirement, UnparsedRequirement]] = None
+            req: Optional[Union[Requirement, UnparsedRequirement]] = None
             known, _ = parser.parse_known_args(line.strip().split())
+
+            # If a requirement is missing hashes but we require them, fail.
+            if not known.hashes and require_hashes:
+                raise PipError(
+                    "invalid: missing hashes for requirement in %s, line %s"
+                    % (filename, lineno)
+                )
+
+            # Similarly, fail if a requirement has hashes but every requirement
+            # we've parsed previously hasn't had them.
+            if known.hashes and not require_hashes and len(name_to_req) > 0:
+                raise PipError(
+                    "invalid: missing hashes for requirements prior to %s, line %s"
+                    % (filename, lineno)
+                )
```

Comment:
N.B.: I figured that these two in-line conditions were better than doing a pass over the entire `name_to_req` dict after-the-fact, both for performance and reasonable exception messages (filename and line context).